src/gt4py/next/program_processors/runners/dace_iterator/itir_to_tasklet.py

```diff
@@ -566,37 +567,117 @@ def builtin_can_deref(
-        (a[0] if isinstance(a, list) else a, b[0] if isinstance(b, list) else b)
-        for a, b in zip(args[1], args[2])
-    ]
-
-    # in case of tuple arguments, generate one if-tasklet for each element of the output tuple
-    if_expr_values = []
-    for a, b in if_expr_args:
-        assert a.dtype == b.dtype
-        expr_args = [
-            (arg, f"{arg.value.data}_v")
-            for arg in (if_node, a, b)
-            if not isinstance(arg, SymbolExpr)
-        ]
-        internals = [
-            arg.value if isinstance(arg, SymbolExpr) else f"{arg.value.data}_v"
-            for arg in (if_node, a, b)
-        ]
-        expr = "({1} if {0} else {2})".format(*internals)
-        if_expr = transformer.add_expr_tasklet(expr_args, expr, a.dtype, "if", dace_debuginfo=di)
-        if_expr_values.append(if_expr[0])
+    assert len(node_args) == 3
+    sdfg = transformer.context.body
+    current_state = transformer.context.state
+    is_start_state = sdfg.start_block == current_state
+
+    # build an empty state to join true and false branches
+    join_state = sdfg.add_state_before(current_state, "join")
+
+    def build_if_state(arg, state):
+        symbol_map = copy.deepcopy(transformer.context.symbol_map)
+        node_context = Context(sdfg, state, symbol_map)
+        node_taskgen = PythonTaskletCodegen(
+            transformer.offset_provider, node_context, transformer.node_types
+        )
+        return node_taskgen.visit(arg)
+
+    # represent the if-statement condition as a tasklet inside an `if_statement` state preceding `join` state
+    stmt_state = sdfg.add_state_before(join_state, "if_statement", is_start_state)
+    stmt_node = build_if_state(node_args[0], stmt_state)[0]
+    assert isinstance(stmt_node, ValueExpr)
+    assert stmt_node.dtype == dace.dtypes.bool
+    assert sdfg.arrays[stmt_node.value.data].shape == (1,)
+
+    # visit true and false branches (here called `tbr` and `fbr`) as separate states, following `if_statement` state
+    tbr_state = sdfg.add_state("true_branch")
+    sdfg.add_edge(
+        stmt_state, tbr_state, dace.InterstateEdge(condition=f"{stmt_node.value.data} == True")
+    )
+    sdfg.add_edge(tbr_state, join_state, dace.InterstateEdge())
+    tbr_values = build_if_state(node_args[1], tbr_state)
+    #
+    fbr_state = sdfg.add_state("false_branch")
+    sdfg.add_edge(
+        stmt_state, fbr_state, dace.InterstateEdge(condition=f"{stmt_node.value.data} == False")
+    )
+    sdfg.add_edge(fbr_state, join_state, dace.InterstateEdge())
+    fbr_values = build_if_state(node_args[2], fbr_state)
+
+    assert isinstance(stmt_node, ValueExpr)
+    assert stmt_node.dtype == dace.dtypes.bool
+    # make the result of the if-statement evaluation available inside current state
+    ctx_stmt_node = ValueExpr(current_state.add_access(stmt_node.value.data), stmt_node.dtype)
+
+    # we distinguish between select if-statements, where both true and false branches are symbolic expressions,
```

Comment:
Good point, but based on DaCe documentation symbols are supposed to remain constant within the state scope and only change on inter-state edges.

> A particular reason that makes symbols useful is the fact they stay constant throughout their defined scope. A symbol defined in a scope (e.g., map parameter) cannot change at all, and symbols that are defined outside an SDFG state cannot be modified inside a state, only in assignments of state transitions.

I assume it should be DaCe's responsibility during lowering to C++/CUDA code to ensure that symbolic expressions are evaluated and assigned to scope variables before entering the code-block of each state.

Reply:
What I was saying is, that if you do an invalid memory access, it does not mater if it happens in a symbol assignment expression or in a Tasklet.
If `idx` is an index that is out of bound (for array `array`) the expression `array[idx]` will be invalid regardless if it happens in a symbol access or not.
Here is an example where an invalid access happens:
```
sdfg = dace.SDFG("Test")
sdfg.add_array("A", shape=(1000,), dtype=dace.float64, transient=False)
sdfg.add_array("__return", shape=(1000,), dtype=dace.float64, transient=False)

init_state = sdfg.add_state("init", is_start_state=True)
comp_state = sdfg.add_state("comp_state")
sdfg.add_edge(
    init_state,
    comp_state,
    dace.InterstateEdge(
        assignments={'invalid_symbol': 'A[999 + 2]'}
    )
)

comp_state.add_mapped_tasklet(
    "invalid_access",
    map_ranges=[('__i0', '0:1000')],
    inputs=dict(__in=dace.Memlet(data='A', subset='__i0')),
    code='__out = __in + invalid_symbol',
    outputs=dict(__out=dace.Memlet(data='__return', subset='__i0')),
    external_edges=True,
)

A = np.ones(1000)
csdfg = sdfg.compile()

res = csdfg(A=A)
```

Reply:
I am not sure if you can do that.
It would boil down to the assumption that a symbolic expression is always valid and a `ValueExpr` is not, where in the code is this assumption established?

I mean it would make sense to establish this guarantee, because it would allow further optimizations, but I do not see that this is currently done.

Reply:
I understood that, though in a second pass :) and I totally agree with you. Based on the way we lower ITIR to SDFG, we do not access arrays on inter-state edges to assign state symbols, so hopefully this is not a problem. Otherwise I have no idea how to ensure that symbol values are the result of valid symbolic expressions.